### ISSUE TITLE
Non-kerchunk backend for GRIB files.

### ISSUE THREAD

**Issue by sharkinsspatial** [zarr-developers/VirtualiZarr#312]
I did a bit of investigation into wrapping the [existing kerchunk grib reader](https://github.com/fsspec/kerchunk/blob/main/kerchunk/grib2.py) to create a GRIB backend but discovered that `kerchunk` [forces inlining of derived coordinates that are not stored at the message level](https://github.com/fsspec/kerchunk/blob/main/kerchunk/grib2.py#L292-L300).  `Virtualizarr` does not support [reading inlined refs](https://github.com/zarr-developers/VirtualiZarr/blob/main/virtualizarr/manifests/manifest.py#L314-L316).  This might be a good reason to kick off work on a `Virtualizarr` specific backend without a direct kerchunk dependency (which was an eventual goal).

I personally have limited experience working with GRIB internals so it would be valuable to get input here from someone with deeper experience like @mpiannucci.  A few questions

1. Should we consider https://github.com/mpiannucci/gribberish for `ChunkManifest` generation (will it see continued development/maintenance)?
2. I'm leaning towards a model where we use `gribberish` as an optional dependency in `Virtualizarr` and place the backend code in this project rather than generating `ChunkManifests` via `gribberish` as is currently done for kerchunk refs via `scan_gribberish` but I don't have strong opinions on this.
3. Can we assume coordinate alignment across all messages in a GRIB file and use `open_virtual_datatree` or should our we also include an `open_virtual_groups` method for problematic datasets?  @mpiannucci we'd also probably need some recommendations for from you about documentation we can include on the types of concat and grouping operations users would perform on the `dict` returned by `open_virtual_groups`.

ref https://github.com/zarr-developers/VirtualiZarr/issues/11 ref https://github.com/zarr-developers/VirtualiZarr/issues/238

**Comment by mpiannucci** [comment]
1. I am happy to support building a gribberish backend for virutalizarr. I personally rely on gribberish for 3 different production apps so while it does not have the cf compliance of cfgrib, I am motivated to improve it tactically. 
2. I would make gribberish optional if you want to use it. 
3. The kerchunk backend for gribberish just flatmaps coordinates which is probably not what a lot of people want (I wanted to avoid data trees)

An alternative is to build a non kerhcunk cf_grib backend or update the kerchunk grib backend to grab the latitude and longitude from the codec. This is [already supported](https://github.com/fsspec/kerchunk/blob/ccf431b6751f4311975087c5798da50cf36753f9/kerchunk/codecs.py#L93) but it is not how kerchunk works as default. 

The reason that is works like this is because GRIB2 encodes coordinates and in many cases the coords are generated from metadata. So if you can, doign it up front and shoving it into bytes is smart. But you can just as easily force generation of the coord data from the codec, it doesnt even matter which grib message you ask for it, they will all be able to do so because every grib message includes all the metadata needed to give back the coordinates. 

To summarize:

1. I think the best case would be to build a simple wrapper around cfgrib to start, because it is the most compliant
2. I support the idea of using gribberish, I can try to help as much as my time allows but i cannot promise anything
3. Either way you choose, you can always use the Grib Codec to get the coordinates instead of always forcing them to be inline. 

I hope this was helpful

**Comment by darothen** [comment]
Following up here, and noticing yesterday's updates - @maxrjones, is there a target date for v2.0? GRIB support here would be a big unblocker for a current project.

**Comment by darothen** [comment]
Hey @sharkinsspatial that's really cool - looking forward to trying that out sometime today!

The two datasets I'm most interested are the open GFS archive on GCS at gs://global-forecast-system (prefixes **gfs.YYYYMMDD/HH/atmos/gfs.tHHz.pgrb2.0p25.fXXX** and the IFS forecasts from the ECMWF open data archive on GCS at gs://ecmwf-open-data (prefixes **YYYYMMDD/HHz/ifs/0p25/oper/YYYYMMDDHHmmss-X-oper-fx.grib2**) where *X* is the forecast lead time in hours. There are other interesting products in the ECMWF open data archive (and it has a more "official" version on S3), but these are simpler for a proof case.

I work with these datasets daily; if there are additional details I can provide or sample code for working with them, let me know! One of our goals is to create a lazily-referenced version of these data archives, subset with (a) the fields required to initialize a typical MLWP forecast like GraphCast or PanguWeather, and (b) the fields required to run [WeatherBench](https://sites.research.google/gr/weatherbench). 

CC @drewbo

**Comment by frazane** [comment]
Hi all, I am following this issue with a lot of interest. You are doing very cool stuff! I took inspiration from @sharkinsspatial data-specific parser and started working on a parser for the operational NWP data from MeteoSwiss. In my case, I opted for [earthkit-data](https://github.com/ecmwf/earthkit-data) to scan the GRIB messages and extract the metadata. What I find quite neat is that we can control how this is done in a declarative way using [xarray profiles](https://earthkit-data.readthedocs.io/en/latest/guide/xarray/overview.html), which allow to control how the xarray object is constructed (in other words, how GRIB metadata is translated to coords, dims, attrs). I believe it might be an option to consider for a parser, especially for 
> creating a more generalized one in the future

Will share more later, or we can have a chat on the slack space (I just joined) if you would like.

**Comment by frazane** [comment]
@TomNicholas here is is https://github.com/MeteoSwiss/icon-ch-vzarr. This is of course a very rough first implementation but the basic functionality is there.

**Comment by TomNicholas** [comment]
@sharkinsspatial I don't think this should be closed? Since we still do not ship a GRIB parser with VirtualiZarr, and there is no general-case one to point people to either? @mpiannucci and I were saying recently that we do still need this...

**Comment by sharkinsspatial** [comment]
@TomNicholas Agreed.  Apologies this got closed by another issue on one of our internal project boards.
